src/sentry/releases/repository.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import datetime
+from logging import Logger, getLogger
+from typing import TYPE_CHECKING, Generator, List, Optional
+
+from django.db.models import F, Q, QuerySet
+
+from sentry.models.deploy import Deploy
+from sentry.models.release import Release
+from sentry.models.releaseprojectenvironment import ReleaseProjectEnvironment
+from sentry.services.hybrid_cloud.organization import RpcOrganization
+
+if TYPE_CHECKING:
+    from sentry.models.organization import Organization
+    from sentry.models.project import Project
+    from sentry.models.release_threshold.release_threshold import ReleaseThreshold
+
+default_logger = getLogger("sentry.releases.repository")
+
+
+@dataclass
+class TimeRange:
+    start: datetime
+    end: datetime
+
+
+@dataclass
+class ReleaseThresholds:
+    release: Release
+    project: Project
+    thresholds: List[ReleaseThreshold]
+
+    def get_latest_deploy_id_by_threshold(self, index: int) -> Optional[Deploy]:
+        if index >= len(self.thresholds):
+            return None
+
+        threshold = self.thresholds[index]
+        if not threshold.environment:
+            return None
+
+        # NOTE: if a threshold has no environment set, we monitor from start of the release creation
+        # If a deploy does not exist for the thresholds environment, we monitor from start of release creation
+        # ReleaseProjectEnvironment model
+        rpe_entry: ReleaseProjectEnvironment | None = next(
+            (
+                rpe
+                for rpe in self.release.releaseprojectenvironment_set.all()
+                if rpe.environment == threshold.environment and rpe.project == self.project
+            ),
+            None,
+        )
+        if not rpe_entry:
+            return None
+
+        last_deploy_id = rpe_entry.last_deploy_id
+        latest_deploy = next(
+            (deploy for deploy in self.release.deploy_set.all() if deploy.id == last_deploy_id),
+            None,
+        )
+
+        return latest_deploy
+
+    def get_threshold_key(self) -> str:
+        """
+        Consistent key helps to determine which thresholds can be grouped together.
+        project_slug - release_version
+
+        NOTE: release versions can contain special characters... `-` delimiter may not be appropriate
+        """
+        return f"{self.project.slug}-{self.release.version}"
+
+
+class ReleaseThresholdsRepository:
```

Comment:
From what I've gathered, a "threshold" is not independent to just the "release" model. I would agree that if a threshold was a singular domain within release, that would be best, but it's actually an accumulation of multiple different models and domains. If it were truly independent, we would be able to do this query backwards, and for each threshold, get all the proper data (i.e. go from threshold, and get the project and release it is attached to, as opposed to the other way around of how we're currently doing it)

Reply:
If we want to move this to the release model, we should be able to produce some sort of domain object that is flat to this domain, with the references to it's dependent, within a query. So something like:
```
class ReleaseThreshold:
    project_id: int
    release_id: int
    deploy_id: int
    name: str
```

Reply:
So actually you can check out the Threshold model to see it's ForeignKey relations

The Threshold is actually tied to a Project
A Project is tied to a Release

So thresholds actually _are_ independent of releases which is why we need to do this roundabout calculations to determine which projects belong to this release, and then which thresholds belong to those projects.

Reply:
> we definitely could query backwards, but that's not very useful for this scenario
Isn't that what we're achieving with the grouping of thresholds with a specific release and project though?

Reply:
> So thresholds actually are independent of releases which is why we need to do this roundabout calculations to determine which projects belong to this release, and then which thresholds belong to those projects.

My main thing is, why can't we do this with joins and a where/in clause? Why the entire bulk of objects for all 3 domains (threshold, release, project)

Reply:
hmmm... Maybe i might be misunderstanding your question here.

Within this `ReleaseThresholdRepository` class, we have 2 methods:
- get releases
- get thresholds

So what i'm suggesting is that the `get_releases` filter can be moved into the release model manager. This should be fairly straight forwards since we're just filtering given a list of release/projects/environments - not entirely necessary since it's just an abstraction around djangos `.filter`.

`get_thresholds` needs to do a bit more gymnastics, but ultimately what this is doing is getting the relevant thresholds _for the specified release_

Since releases can have multiple projects and projects can have multiple thresholds, this the chain we're going through.
So - maybe instead of adding this helper to the manager, we'd add it to the Release model itself so you could do something like `Release.get_thresholds()`

> why can't we do this with joins and a where/in clause

hmmmm - that's a good call. for grabbing all the thresholds, we probably _could_ just do this via SQL joining all 3 models? but this repository class still doesn't help with that right?